File: src/spec/load.ts
/** Load a graph spec from YAML/JSON, validate it, resolve subgraph files, and apply defaults. */
import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";
import { GraphSpecZ, type GraphSpec, type NodeSpec } from "./schema.js";

export interface LoadedGraph {
  spec: GraphSpec;
  /** Absolute path of the spec file (or "<inline>"). */
  file: string;
  /** Directory used to resolve `module:` and `graph:` paths. */
  dir: string;
}

export class SpecError extends Error {
  constructor(
    message: string,
    public issues: string[] = [],
  ) {
    super(message);
  }
}

export function parseSpecText(text: string, file = "<inline>"): GraphSpec {
  let raw: unknown;
  try {
    raw = YAML.parse(text);
  } catch (e) {
    throw new SpecError(`${file}: could not parse YAML/JSON: ${(e as Error).message}`);
  }
  return validateSpecObject(raw, file);
}

export function validateSpecObject(raw: unknown, file = "<inline>"): GraphSpec {
  const res = GraphSpecZ.safeParse(raw);
  if (!res.success) {
    const issues = res.error.issues.map((i) => `${i.path.join(".") || "<root>"}: ${i.message}`);
    throw new SpecError(`${file}: invalid graph spec (${issues.length} issue${issues.length === 1 ? "" : "s"})`, issues);
  }
  return res.data;
}

export function loadGraph(file: string, seen: Set<string> = new Set()): LoadedGraph {
  const abs = path.resolve(file);
  if (seen.has(abs)) throw new SpecError(`subgraph cycle: ${abs} includes itself`);
  seen.add(abs);
  if (!fs.existsSync(abs)) throw new SpecError(`spec file not found: ${abs}`);
  const text = fs.readFileSync(abs, "utf8");
  const spec = parseSpecText(text, abs);
  const dir = path.dirname(abs);
  resolveNested(spec, dir, seen);
  return { spec, file: abs, dir };
}

export function loadGraphFromObject(raw: unknown, dir = process.cwd()): LoadedGraph {
  const spec = validateSpecObject(raw);
  resolveNested(spec, dir, new Set());
  return { spec, file: "<inline>", dir };
}

/** Inline `graph: ./file.yaml` subgraphs and make module paths absolute so nested runs don't depend on cwd. */
function resolveNested(spec: GraphSpec, dir: string, seen: Set<string>) {
  for (const node of spec.nodes as NodeSpec[]) {
    if (node.kind === "subgraph" && typeof node.graph === "string") {
      const sub = loadGraph(path.resolve(dir, node.graph), new Set(seen));
      node.graph = sub.spec;
    } else if (node.kind === "subgraph") {
      resolveNested(node.graph as GraphSpec, dir, seen);
    }
    if (node.kind === "loop") resolveNested(node.body, dir, seen);
    if ((node.kind === "code" || node.kind === "verify") && node.module && !path.isAbsolute(node.module)) {
      node.module = path.resolve(dir, node.module);
    }
  }
}

export function nodeById(spec: GraphSpec, id: string): NodeSpec | undefined {
  return spec.nodes.find((n) => n.id === id);
}

File: gren/spec/load.py
"""Load a graph spec from YAML/JSON, validate it, resolve subgraph files, make module paths absolute."""
from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import yaml
from pydantic import ValidationError

from .schema import GraphSpec


class SpecError(Exception):
    def __init__(self, message: str, issues: list[str] | None = None):
        super().__init__(message)
        self.message = message
        self.issues = issues or []


@dataclass
class LoadedGraph:
    spec: GraphSpec
    file: str
    dir: str


def validate_spec_object(raw: Any, file: str = "<inline>") -> GraphSpec:
    try:
        return GraphSpec.model_validate(raw)
    except ValidationError as e:
        issues = []
        for err in e.errors():
            loc = ".".join(str(p) for p in err.get("loc", ())) or "<root>"
            issues.append(f"{loc}: {err.get('msg')}")
        raise SpecError(f"{file}: invalid graph spec ({len(issues)} issue{'s' if len(issues) != 1 else ''})", issues) from None


def parse_spec_text(text: str, file: str = "<inline>") -> GraphSpec:
    try:
        raw = yaml.safe_load(text)
    except yaml.YAMLError as e:
        raise SpecError(f"{file}: could not parse YAML/JSON: {e}") from None
    return validate_spec_object(raw, file)


def _resolve_nested(spec: GraphSpec, base_dir: str, seen: set[str]) -> None:
    for node in spec.nodes:
        kind = node.kind
        if kind == "subgraph":
            if isinstance(node.graph, str):
                sub = load_graph(os.path.join(base_dir, node.graph), set(seen))
                node.graph = sub.spec
            else:
                _resolve_nested(node.graph, base_dir, seen)
        elif kind == "loop":
            _resolve_nested(node.body, base_dir, seen)
        if kind in ("code", "verify") and getattr(node, "module", None) and not os.path.isabs(node.module):
            node.module = os.path.abspath(os.path.join(base_dir, node.module))


def load_graph(file: str, seen: set[str] | None = None) -> LoadedGraph:
    seen = seen if seen is not None else set()
    abs_path = os.path.abspath(file)
    if abs_path in seen:
        raise SpecError(f"subgraph cycle: {abs_path} includes itself")
    seen.add(abs_path)
    if not os.path.exists(abs_path):
        raise SpecError(f"spec file not found: {abs_path}")
    text = Path(abs_path).read_text(encoding="utf8")
    spec = parse_spec_text(text, abs_path)
    base_dir = os.path.dirname(abs_path)
    _resolve_nested(spec, base_dir, seen)
    return LoadedGraph(spec=spec, file=abs_path, dir=base_dir)


def load_graph_from_object(raw: Any, base_dir: str | None = None) -> LoadedGraph:
    spec = validate_spec_object(raw)
    base_dir = base_dir or os.getcwd()
    _resolve_nested(spec, base_dir, set())
    return LoadedGraph(spec=spec, file="<inline>", dir=base_dir)
